### PR TITLE
chore(config): initial integration of configuration schema for sinks (part one)

### PR DESCRIPTION
**Pull request by tobz**
(part of the work required for https://github.com/vectordotdev/vector/issues/12223)

This PR is the initial implementation of the configuration schema framework, specifically for sinks. It represents the first major chunk of conversion work for sinks. Frankly, there is a large number of sinks, and I wanted to get some of the work merged in sooner rather than later, and to make the reviewing a little less terrible.

It does not change how we load configurations, or anything about our use of `inventory`/`typetag`/`SourceOuter`/etc. That work will be the final step after all components have been minimally instrumented/converted to support having their configuration schema generated.

# Reviewer Notes

In general, the majority of this PR is indeed the instrumenting of many sinks and porting over the Cue documentation to the code. I did have to do some work to the actual `vector-config`/`vector-config-macros` side, but it's reasonably small and I'll add inline comments on this PR to explain them a little better.

Other than that, I've strived to port over the existing Cue documentation in a faithful manner, but there are a number of small divergences:

- making the documentation of common fields consistent across transforms
- fixing spelling/grammatical issues, changing from passive to active voice
- some of the Cue-specific links have stayed as their Cue syntax (because there was no better alternative at the moment) but I've switched to normal Markdown links where possible

### CODE REVIEW

File: lib/vector-config-common/src/validation.rs [hunk 1]
```diff
@@ -261,7 +261,7 @@ impl ToTokens for Validation {
     fn to_tokens(&self, tokens: &mut TokenStream) {
         let validation_tokens = match self {
             Validation::KnownFormat(format) => {
-                quote! { ::vector_config::validation::Validation::Format(#format) }
+                quote! { ::vector_config::validation::Validation::KnownFormat(#format) }
```
Comment (tobz) [inline lib/vector-config-common/src/validation.rs:264]: Simple bug fix, as we were emitting an invalid enum variant here.

File: lib/vector-config-macros/src/ast/field.rs [hunk 2]
```diff
@@ -89,6 +90,15 @@ impl<'a> Field<'a> {
...
 
+impl<'a> Spanned for Field<'a> {
+    fn span(&self) -> proc_macro2::Span {
+        match self.original.ident.as_ref() {
+            Some(ident) => ident.span(),
+            None => self.original.ty.span(),
+        }
+    }
+}
+
```
Comment (tobz) [inline lib/vector-config-macros/src/ast/field.rs:101]: This change is us trying to generate better compiler error output from the procedural macro, specifically: for named fields (`foo: String`) we underline the field name when the macro encounters an error, and if it's a tuple field (the string in `struct OuterStruct(String)`), we underline the type itself.

Prior to this, the spanning (how the compiler tracks tokens for the purposes of ensuring errors end up underlining the right code in error output) wasn't up to snuff so errors weren't being attached as narrowly as we could actually achieve.

File: lib/vector-config-macros/src/configurable.rs [hunk 3]
```diff
@@ -136,14 +136,18 @@ fn build_struct_generate_schema_fn(
...
     let field_metadata = generate_field_metadata(&field_metadata_ref, field);
 
+    let spanned_generate_schema = quote_spanned! {field.span()=>
+        <#field_ty as ::vector_config::Configurable>::generate_schema(schema_gen, #field_metadata_ref.as_subschema())
+    };
+
     quote! {
         #field_metadata
-        let mut subschema = #field_as_configurable::generate_schema(schema_gen, #field_metadata_ref.as_subschema());
+        let mut subschema = #spanned_generate_schema;
```
Comment (tobz) [inline lib/vector-config-macros/src/configurable.rs:150]: Related to the "better code underlining in errors" stuff.

File: lib/vector-config-macros/src/configurable.rs [hunk 4]
```diff
@@ -177,10 +181,13 @@ fn generate_named_struct_field(
...
         // field is a part of, then we consider it required unless the field type itself is inherently
         // optional, such as being `Option<T>`.
+        let spanned_is_optional = quote_spanned! {field.span()=>
+            <#field_ty as ::vector_config::Configurable>::is_optional()
+        };
         let maybe_field_required =
             if container.default_value().is_none() && field.default_value().is_none() {
                 Some(quote! {
-                    if !#field_as_configurable::is_optional() {
+                    if !#spanned_is_optional {
```
Comment (tobz) [inline lib/vector-config-macros/src/configurable.rs:190]: Related to the "better code underlining in errors" stuff.

File: lib/vector-config-macros/src/configurable.rs [hunk 5]
```diff
@@ -1,7 +1,7 @@
 use proc_macro::TokenStream;
 use proc_macro2::Span;
-use quote::quote;
-use syn::{parse_macro_input, parse_quote, DeriveInput, ExprPath, Ident, Type};
+use quote::{quote, quote_spanned};
```
Comment (tobz) [inline lib/vector-config-macros/src/configurable.rs:3]: Any of the changes in this file that have the term "span" or "spanned" in them are related to the "better code underlining in errors" stuff.

File: src/sinks/redis.rs [hunk 6]
```diff
@@ -85,20 +113,42 @@ impl SinkBatchSettings for RedisDefaultBatchSettings {
...
+    ///
+    /// The URL _must_ take the form of `protocol://server:port/db` where the protocol can either be
+    /// `redis` or `rediss` for connections secured via TLS.
     url: String,
-    key: String,
+
+    /// The Redis key to publish messages to.
+    #[configurable(metadata(templateable))]
+    #[configurable(validation(length(min = 1)))]
+    key: Template,
```
Comment (tobz) [inline src/sinks/redis.rs:141]: Made this `Template` because it got parsed as a `Template` during build, and we can discover any errors with the input string way earlier up the stack... and eventually, we'd need to do this anyways when we want to encode the `Template` allowable format, etc, into the configuration schema.

File: lib/vector-config-macros/src/ast/field.rs [hunk 7]
```diff
@@ -23,6 +23,7 @@ impl<'a> Field<'a> {
         is_virtual_newtype: bool,
     ) -> darling::Result<Field<'a>> {
         let original = serde.original;
+
```
Comment (spencergilbert) [inline lib/vector-config-macros/src/ast/field.rs:26]: Do you have a good rule-of-thumb for line breaking (assuming this was you and not auto-fmting). I agree it looks better, but often feel odd about adding breaks whenever I feel like it.

File: lib/vector-config-macros/src/configurable.rs [hunk 8]
```diff
@@ -350,36 +360,25 @@ fn generate_variant_metadata(
...
     let maybe_title = get_metadata_title(meta_ident, variant.title());
     let description = get_metadata_description(meta_ident, variant.description())
         .expect("enum variants without a description should be rejected during AST parsing");
     let maybe_deprecated = get_metadata_deprecated(meta_ident, variant.deprecated());
     let maybe_custom_attributes = get_metadata_custom_attributes(meta_ident, variant.metadata());
 
+    // We ase a unit tuple here because we don't have access to the actual type of the variant's enum container,
+    // at least not without parsing it specifically as a type token from an amalgamated version of the enum ident... but
+    // it doesn't actually matter because enums don't support default values, which is the only spot in `Metadata<T>`
+    // where the `T` gets used, so we can carry all the other pertinent details with `()`... it's still a littlw wonky, though.
```
Comment (spencergilbert) [inline lib/vector-config-macros/src/configurable.rs:372]:
```suggestion
    // We use a unit tuple here because we don't have access to the actual type of the variant's enum container,
    // at least not without parsing it specifically as a type token from an amalgamated version of the enum ident... but
    // it doesn't actually matter because enums don't support default values, which is the only spot in `Metadata<T>`
    // where the `T` gets used, so we can carry all the other pertinent details with `()`... it's still a little wonky, though.
```

"amalgamated" might be a challenging word that we could sub for something simpler.

File: lib/vector-config-macros/src/configurable.rs [hunk 9]
```diff
@@ -350,36 +360,25 @@ fn generate_variant_metadata(
...
-}
-
-fn get_variant_type_as_configurable() -> proc_macro2::TokenStream {
-    // We hardcode this to a unit tuple because we don't have access to the actual type of the variant's enum container,
-    // at least not without parsing it specifically as a type token from an amalgamated version of the enum ident... but
-    // it doesn't actually matter because enums don't support default values, which is the only spot in `Metadata<T>`
-    // where the `T` gets used, so we can carry all the other pertinent details with `()`... it's still a littlw wonky, though.
-    quote! { <() as ::vector_config::Configurable> }
-}
-
```
Comment (spencergilbert) [inline lib/vector-config-macros/src/configurable.rs:382]: I'm curious why you moved from these functions to what appears to be their original output in the final `quote!` macros.

File: lib/vector-config/src/lib.rs [hunk 10]
```diff
@@ -81,6 +81,32 @@
 // can essentially define flags i.e. `docs:templateable` as a metadata value for marking a field as working with
 // Vector's template syntax, since doing `templateable = true` is weird given that we never otherwise specifically
 // disable it. In other words, we want a way to define feature flags in metadata.
+//
+// TODO: Should we add a way, and/or make it the default, that if you only supply a description of a
+// field, it concats the description of the type of the field? for example, you have:
```
Comment (spencergilbert) [inline lib/vector-config/src/lib.rs:86]: 👍 does feel like a nice addition.

File: lib/vector-config/src/stdlib.rs [hunk 11]
```diff
@@ -115,12 +115,41 @@ where
...
+{
+    fn is_optional() -> bool {
+        // A map with required fields would be... an object.  So if you want that, make a struct
+        // instead, not a hashmap.
+        true
+    }
+
+    fn generate_schema(gen: &mut SchemaGenerator, overrides: Metadata<Self>) -> SchemaObject {
+        // We explicitly do not pass anything from the override metadata, because there's nothing to
+        // reasonably pass: if `V` is referencable, using the description for `BTreeMap<String, V>`
```
Comment (spencergilbert) [inline lib/vector-config/src/stdlib.rs:130]:
```suggestion
        // reasonably pass: if `V` is referenceable, using the description for `BTreeMap<String, V>`
```

Chrome still says it's misspelled but 🤷

File: src/sinks/aws_s3/config.rs [hunk 12]
```diff
@@ -35,29 +35,83 @@ const DEFAULT_KEY_PREFIX: &str = "date=%F/";
...
 
-#[derive(Deserialize, Serialize, Debug, Clone)]
-#[serde(deny_unknown_fields)]
+/// Configuration for the `aws_s3` sink.
+#[configurable_component(sink)]
+#[derive(Clone, Debug)]
 pub struct S3SinkConfig {
+    /// The S3 bucket name.
+    ///
+    /// This must not include a leading `s3://` or a trailing `/`.
```
Comment (spencergilbert) [inline src/sinks/aws_s3/config.rs:44]: A thought on consistency, which you'd mentioned offline, I remember seeing both `must` and `_must_`. Is this something we'd like to add to any "writing guidelines" we end up having?

File: src/sinks/aws_s3/config.rs [hunk 13]
```diff
@@ -35,29 +35,83 @@ const DEFAULT_KEY_PREFIX: &str = "date=%F/";
...
     pub filename_extension: Option<String>,
+
     #[serde(flatten)]
     pub options: S3Options,
+
     #[serde(flatten)]
     pub region: RegionOrEndpoint,
+
     #[serde(flatten)]
     pub encoding: EncodingConfigWithFraming,
```
Comment (spencergilbert) [inline src/sinks/aws_s3/config.rs:93]: I would have expected these to need a `#[configurable(derived)]`?

File: src/sinks/azure_monitor_logs.rs [hunk 14]
```diff
@@ -32,25 +33,59 @@ fn default_host() -> String {
...
+    ///
+    /// [shared_key]: https://docs.microsoft.com/en-us/azure/azure-monitor/platform/data-collector-api#authorization
     pub shared_key: String,
+
+    /// The [record type][record_type] of the data that is being submitted.
+    ///
+    /// Can only contain letters, numbers, and underscores (_), and may not exceed 100 characters.
+    ///
+    /// [record_type]: https://docs.microsoft.com/en-us/azure/azure-monitor/platform/data-collector-api#request-headers
+    #[configurable(validation(pattern = "[a-zA-Z0-9_]{1,100}"))]
```
Comment (spencergilbert) [inline src/sinks/azure_monitor_logs.rs:56]: YES! I love this. ❤️

File: src/sinks/datadog/logs/config.rs [hunk 15]
```diff
@@ -45,33 +45,55 @@ impl SinkBatchSettings for DatadogLogsDefaultBatchSettings {
...
+#[derive(Clone, Debug, Default)]
 #[serde(deny_unknown_fields)]
-pub(crate) struct DatadogLogsConfig {
+pub struct DatadogLogsConfig {
+    /// The endpoint to send logs to.
     pub(crate) endpoint: Option<String>,
-    // Deprecated, replaced by the site option
+
+    /// The Datadog region to send logs to.
+    #[configurable(deprecated)]
```
Comment (spencergilbert) [inline src/sinks/datadog/logs/config.rs:57]: Should we still be noting what has superseded the deprecated option?

File: src/sinks/datadog_archives.rs [hunk 16]
```diff
@@ -115,43 +140,112 @@ pub struct DatadogArchivesSinkConfig {
     acknowledgements: AcknowledgementsConfig,
 }
 
-#[derive(Deserialize, Serialize, Default, Debug, Clone)]
-#[serde(deny_unknown_fields)]
```
Comment (spencergilbert) [inline src/sinks/datadog_archives.rs:119]: Was this intentionally dropped?

File: src/sinks/datadog_archives.rs [hunk 17]
```diff
@@ -115,43 +140,112 @@ pub struct DatadogArchivesSinkConfig {
...
 #[serde(deny_unknown_fields)]
 pub struct AzureBlobConfig {
+    /// The Azure Blob Storage Account connection string.
+    ///
+    /// Authentication with access key is the only supported authentication method.
     pub connection_string: String,
 }
 
-#[derive(Deserialize, Serialize, Default, Debug, Clone)]
-#[serde(deny_unknown_fields)]
```
Comment (spencergilbert) [inline src/sinks/datadog_archives.rs:150]: Same as above, was this intentionally dropped?

File: src/sinks/elasticsearch/config.rs [hunk 18]
```diff
@@ -38,41 +38,89 @@ use lookup::path;
...
 
-#[derive(Deserialize, Serialize, Debug, Clone, Default)]
+/// Configuration for the `elasticsearch` sink.
+#[configurable_component(sink)]
+#[derive(Clone, Debug, Default)]
 #[serde(deny_unknown_fields)]
 pub struct ElasticsearchConfig {
+    /// The Elasticsearch endpoint to send logs to.
+    ///
+    /// This should be the full URL as shown in the example.
```
Comment (spencergilbert) [inline src/sinks/elasticsearch/config.rs:48]: 🤔 unrelated to your PR, but I wonder why this isn't a more specific type than String.

File: src/sinks/file/mod.rs [hunk 19]
```diff
@@ -74,10 +86,16 @@ impl GenerateConfig for FileSinkConfig {
     }
 }
 
-#[derive(Deserialize, Serialize, Debug, Eq, PartialEq, Clone, Copy)]
+/// Compression algorithm.
+// TODO: Why doesn't this already use `crate::sinks::util::Compression`? :thinkies:
```
Comment (spencergilbert) [inline src/sinks/file/mod.rs:90]: Older and never updated to use the new utils? Open a tech-debt issue?

File: src/sinks/gcp/cloud_storage.rs [hunk 20]
```diff
@@ -53,28 +53,96 @@ pub enum GcsHealthcheckError {
 
 const NAME: &str = "gcp_cloud_storage";
 
-#[derive(Deserialize, Serialize, Debug)]
-#[serde(deny_unknown_fields)]
```
Comment (spencergilbert) [inline src/sinks/gcp/cloud_storage.rs:57]: Another dropped deny_unknown...

File: src/sinks/gcp/stackdriver_logs.rs [hunk 21]
```diff
@@ -33,34 +33,63 @@ enum HealthcheckError {
...
+    #[configurable(metadata(templateable))]
     pub log_id: Template,
 
+    /// The monitored resource to associate the logs with.
     pub resource: StackdriverResource,
+
+    /// The field of the log event from which to take the outgoing log’s `severity` field.
+    ///
+    /// The named field is removed from the log event if present, and must be either an integer
+    /// between 0 and 800 or a string containing one of the [severity level names][sev_names] (case
```
Comment (spencergilbert) [inline src/sinks/gcp/stackdriver_logs.rs:59]: 👀 wth is a severity of 800?!

File: src/sinks/gcp/stackdriver_logs.rs [hunk 22]
```diff
@@ -84,24 +113,61 @@ struct StackdriverSink {
...
+/// Monitored resources in GCP allow associating logs and metrics specifically with native resources
+/// within Google Cloud Platform. This takes the form of a "type" field which identifies the
+/// resource, and a set of type-specific labels to uniquely identify a resource of that type.
+///
+/// See [Monitored resource types][mon_docs] for more information.
+///
+/// [mon_docs]: https://cloud.google.com/monitoring/api/resources
+// TODO: this type is specific to the stackdrivers log sink because it allows for template-able
+// label values, but we should consider replacing `sinks::gcp::GcpTypedResource` with this so both
+// the stackdriver metrics _and_ logs sink can have template-able label values, and less duplication
```
Comment (spencergilbert) [inline src/sinks/gcp/stackdriver_logs.rs:160]: 💯 I had started playing around with this + updating the traits for these sinks.

File: src/sinks/humio/logs.rs [hunk 23]
```diff
@@ -25,37 +25,98 @@ use crate::{
...
 pub struct HumioLogsConfig {
+    /// The Humio ingestion token.
     pub(super) token: String,
-    // Deprecated name
+
+    /// The base URL of the Humio instance.
     #[serde(alias = "host")]
     pub(super) endpoint: Option<String>,
+
+    /// The source of events sent to this sink.
```
Comment (spencergilbert) [inline src/sinks/humio/logs.rs:40]: I don't have a suggestion, but I'm wondering if this could be confusing with Vector's internal use of "source" as well.

File: src/sinks/mod.rs [hunk 24]
```diff
@@ -116,7 +117,6 @@ pub enum HealthcheckError {
     UnexpectedStatus { status: ::http::StatusCode },
 }
 
-/*
```
Comment (spencergilbert) [inline src/sinks/mod.rs:119]: Number of stray block comment pieces floating around in this file

File: src/sinks/redis.rs [hunk 25]
```diff
@@ -85,20 +113,42 @@ impl SinkBatchSettings for RedisDefaultBatchSettings {
     const TIMEOUT_SECS: f64 = 1.0;
 }
 
-#[derive(Clone, Debug, Deserialize, Serialize)]
+/// Configuration for the `redis`gsink.
```
Comment (spencergilbert) [inline src/sinks/redis.rs:116]:
```suggestion
/// Configuration for the `redis` sink.
```

File: lib/vector-config-macros/src/ast/field.rs (hunk 7 above)
Comment (tobz) [reply lib/vector-config-macros/src/ast/field.rs:26]: Hmm, my rule-of-thumb is a bit subjective but: for functions, I typically group variables that are common/needed for the rest of the function body at the top, and then try to use line breaks to group up individual steps:

- alright, convert this input to the intended format
- alright, now extract the data we need
- alright, for all the data we extracted, transform it
- alright, now re-package the transformed data and return it
- etc

File: lib/vector-config-macros/src/configurable.rs (hunk 9 above)
Comment (tobz) [reply lib/vector-config-macros/src/configurable.rs:382]: This got removed because I couldn't actually get it to generate the correct spanned output and had to group everything together in a single `quote_spanned!` rather than doing it piecemeal.

Essentially, I tried to do all of the `quote_spanned!` stuff in these functions first, and use them in an interpolated fashion, but it just... didn't work.

File: lib/vector-config/src/lib.rs (hunk 10 above)
Comment (tobz) [reply lib/vector-config/src/lib.rs:86]: Yeah, it would make it a lot easier to only partially customize the description of a field using a type that already has a rich description.

File: lib/vector-config/src/stdlib.rs (hunk 11 above)
Comment (tobz) [reply lib/vector-config/src/stdlib.rs:130]: Yeah, it's not a word in the English dictionary but it's common enough in software that I decided to use it. I'm gonna change all instances of `referencable` to `referenceable` for clarity, though.

File: src/sinks/aws_s3/config.rs (hunk 12 above)
Comment (tobz) [reply src/sinks/aws_s3/config.rs:44]: Hmm, probably. Haven't even really gotten into worrying too much about the styling of the doc comments.

File: src/sinks/aws_s3/config.rs (hunk 13 above)
Comment (tobz) [reply src/sinks/aws_s3/config.rs:93]: When fields are flattened, we're replacing this one field with N fields in the actual configuration file, so we don't need a title/description for `options` or `S3Options`... so long as each of the fields on `S3Options` have a title/description (or are derived, etc...).

File: src/sinks/azure_monitor_logs.rs (hunk 14 above)
Comment (tobz) [reply src/sinks/azure_monitor_logs.rs:56]: Yeah! It'll be so much nicer to start baking validation in to the struct definition.

I also have plans to make the macro compile any `pattern` validators with the `regex` crate to tell the user if the regular expression itself is actually valid or not. 😈

File: src/sinks/datadog/logs/config.rs (hunk 15 above)
Comment (tobz) [reply src/sinks/datadog/logs/config.rs:57]: Hmm, probably.

My intent is to add support to the `deprecated` helper attribute that lets you specify the name of another field on the object that replaces it... but it has some edge cases (like if the field only exists via a `#[serde(flatten)]`-ed field, since ideally we would want to enforce that the field you're saying supersedes it actually exists.

The flipside is that the intent is to also use `#[serde(alias = "...")]` as a signal for deprecated fields, since we should never allow two identical ways to set a field save for backwards compatibility, and a backwards compatible field implies that it is deprecated.

So I'm not sure, we do need a way to support both, because if the field is an entirely different type, you can't just alias it, but it would also be nice to have a consistent way to mark/demonstrate a field being deprecated.

Long story short, I'll update the comment to refer to the field that supersedes it.

File: src/sinks/datadog_archives.rs (hunk 16 above)
Comment (tobz) [reply src/sinks/datadog_archives.rs:119]: Actually, eh, I did intend to drop it but thinking through, this is still a valid use of it.

File: src/sinks/elasticsearch/config.rs (hunk 18 above)
Comment (tobz) [reply src/sinks/elasticsearch/config.rs:48]: It should be `UriSerde` or similar, but I decided to leave it as-is for the moment.

File: lib/vector-config/src/stdlib.rs (hunk 11 above)
Comment (spencergilbert) [reply lib/vector-config/src/stdlib.rs:130]: It's definitely what gets returned by googling so seems good enough 😄

File: src/sinks/aws_s3/config.rs (hunk 12 above)
Comment (spencergilbert) [reply src/sinks/aws_s3/config.rs:44]: Yeah, wasn't a major thing - was just top of mind based on your parting thoughts last night. Most of the comments do appear to be using `_must_`, if memory serves.

File: src/sinks/aws_s3/config.rs (hunk 13 above)
Comment (spencergilbert) [reply src/sinks/aws_s3/config.rs:93]: AH, doh - yeah that makes sense. Silly in hindsight, thanks for explanation.

File: src/sinks/datadog/logs/config.rs (hunk 15 above)
Comment (spencergilbert) [reply src/sinks/datadog/logs/config.rs:57]: Yeah, it seemed like `alias` was a "better" way of doing it, assuming it was a name change and not changing the provided contents as well. I can see how it'd be tricky to make work universally.

Not knowing what you need to replace your deprecated option with is definitely a bad time.

File: src/sinks/elasticsearch/config.rs (hunk 18 above)
Comment (spencergilbert) [reply src/sinks/elasticsearch/config.rs:48]: @jszwedko could we call this out as something to check as part of the integrations audit?

File: src/sinks/redis.rs (hunk 25 above)
Comment (tobz) [reply src/sinks/redis.rs:116]: Good catch. Will fix.

File: src/sinks/mod.rs (hunk 24 above)
Comment (tobz) [reply src/sinks/mod.rs:119]: Yeah, I specifically commented out sinks that I hadn't yet converted, but didn't want to group them all together and lose my alphabetical ordering. :P

File: src/sinks/gcp/stackdriver_logs.rs (hunk 21 above)
Comment (tobz) [reply src/sinks/gcp/stackdriver_logs.rs:59]: Oh damn. 😂 Let me double check my copy/paste job...

Comment (tobz) [reply src/sinks/gcp/stackdriver_logs.rs:59]: Huh, apparently it's [real](https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry#logseverity): each level maps to N*100, so 100, 200, 300.

I guess maybe they wanted to make it possible to allow ordering or something similar to priority levels for each severity level? 🤷🏻

File: src/sinks/datadog_archives.rs (hunk 17 above)
Comment (tobz) [reply src/sinks/datadog_archives.rs:150]: Intentionally, yes, but erroneously. I'll fix that.

File: src/sinks/gcp/cloud_storage.rs (hunk 20 above)
Comment (tobz) [reply src/sinks/gcp/cloud_storage.rs:57]: Will fix.

File: src/sinks/humio/logs.rs (hunk 23 above)
Comment (tobz) [reply src/sinks/humio/logs.rs:40]: Hmmm, fair point, but yeah, not sure what a better phrase would be given that the field in Humio is literally called `@source`.

File: src/sinks/redis.rs (hunk 25 above)
Comment (tobz) [reply src/sinks/redis.rs]: Fixed in 18c22d4b8.

File: lib/vector-config/src/stdlib.rs (hunk 11 above)
Comment (tobz) [reply lib/vector-config/src/stdlib.rs]: Fixed in 18c22d4b8.

File: lib/vector-config-macros/src/configurable.rs (hunk 8 above)
Comment (tobz) [reply lib/vector-config-macros/src/configurable.rs]: Fixed in 18c22d4b8.

File: src/sinks/datadog_archives.rs (hunk 16 above)
Comment (tobz) [reply src/sinks/datadog_archives.rs:119]: Fixed in 18c22d4b8.

File: src/sinks/datadog_archives.rs (hunk 17 above)
Comment (tobz) [reply src/sinks/datadog_archives.rs:150]: Fixed in 18c22d4b8.

File: src/sinks/gcp/cloud_storage.rs (hunk 20 above)
Comment (tobz) [reply src/sinks/gcp/cloud_storage.rs:57]: Fixed in 18c22d4b8.

File: src/sinks/file/mod.rs (hunk 19 above)
Comment (tobz) [reply src/sinks/file/mod.rs:91]: Fineeeeee: opened https://github.com/vectordotdev/vector/issues/13694. :D

File: src/sinks/datadog/logs/config.rs (hunk 15 above)
Comment (tobz) [reply src/sinks/datadog/logs/config.rs]: Fixed the comment in a3fff3e3d.

File: src/sinks/gcp/stackdriver_logs.rs (hunk 21 above)
Comment (spencergilbert) [reply src/sinks/gcp/stackdriver_logs.rs:59]: Yeah, that's just bonkers to me - too many choices!

File: src/sinks/humio/logs.rs (hunk 23 above)
Comment (spencergilbert) [reply src/sinks/humio/logs.rs:40]: Yep 🤷 not a good suggestion or anything to change - I guess there was an old suggestion that we'd be renaming sources/sinks but yeah. Fine as is. Naming is hard.

**Review by spencergilbert — APPROVED** [review]
